Support for custom App Bundle configuration file

mathieubourgeois opened the pull request:
Fixes: https://github.com/xamarin/xamarin-android/issues/3813

Currently, `bundletool` is not configurable in any significant way with Xamarin.Android. However, there are useful properties that can be defined in the configuration file for `bundletool`, especially related to the dimensions on which to split the resulting apks.

Therefore, added a way to set a custom configuration file with `AndroidBundleConfigurationFile`. Xamarin still has final control over the resulting configuration file, and will still set the uncompressed extensions no matter what is set in the configuration file. Knowledge of the underlying configuration is required however, which may be problematic. 

Tested on d16-3 and with it we're able to control the split and the extraction of the native libs.

This was enough to fix our underlying issues for my team, however I'm still open to suggestions on how to adapt it so that it's useful for everyone. I was thinking somebody might want properties for the most common options, which could then be exposed in the IDE, but I was unsure of the best way to proceed from there and still think that having the configuration available to somebody who understands the underlying format is still a useful idea.

### Example `BuildConfig.json`

```
{
    "optimizations": {
        "splits_config": {
            "split_dimension": [
                {
                    "value": "ABI",
                    "negate": false
                },
                {
                    "value": "SCREEN_DENSITY",
                    "negate": true
                },
                {
                    "value": "LANGUAGE",
                    "negate": true
                }
            ]
        },
        "uncompress_native_libraries":
        {
            "enabled":false
        }
    }
}
```

dellis1972 commented on src/Xamarin.Android.Build.Tasks/Tasks/BuildAppBundle.cs:
```diff
@@ -78,13 +81,34 @@ public override bool RunTask ()
+						}
+					}
+					catch (Exception e)
+					{ } // Do nothing for now
```
If there is an exception we should at least log a warning via `Log.LogCodedWarning` with the details of the exception.

jonathanpeppers replied:
You might just let this exception `throw` as we generate a reasonable error code by default now: https://github.com/xamarin/xamarin-android/blob/fbe4b44c7b40dcaacdecaed834c631b15490cc52/src/Xamarin.Android.Build.Tasks/Tasks/AndroidTask.cs#L18-L21

mathieubourgeois replied:
Fixed in 478cf6d

dellis1972 commented:
Thanks for the contribution. @jonathanpeppers is the main developer on the app bundle tooling, so I'll wait for his feedback. But this does seem like a good addition. 

I would ask if we can add some documentation about this new property in https://github.com/xamarin/xamarin-android/blob/master/Documentation/guides/BuildProcess.md#packaging-properties.

mathieubourgeois commented:
Ok great. I'll wait for @jonathanpeppers ' feedback and I will do the requested changes afterwards then.

mathieubourgeois commented on src/Xamarin.Android.Build.Tasks/Tasks/BuildAppBundle.cs:
```diff
@@ -78,13 +81,34 @@ public override bool RunTask ()
-				var json = JsonConvert.SerializeObject (new {
+
+				var json = JObject.FromObject(new { });
+				if(!string.IsNullOrWhiteSpace(CustomBuildConfigFile) && File.Exists(CustomBuildConfigFile))
```
Adjusted (and the rest of that source file) in 478cf6d

mathieubourgeois commented:
Documentation and test added.

For the test, I modified the existing test, as you mentioned. I used the build configuration to stop the aab from splitting along language (i.e. base-en.apk shouldn't appear anymore) and adjusted the test accordingly.

jonathanpeppers commented:
/azp run

mathieubourgeois commented:
The build error doesn't seem related to the pull request (seems like it tried to access a resource at the same time as another process)

dellis1972 commented:
Thanks for the contribution :)

brendanzagaeski commented:
Thanks again for the contribution!

*Release status update*

A new Release version of Xamarin.Android has now been published that includes the new functionality from this item.

New functionality included in Xamarin.Android 10.3.1.0.

New functionality included on Windows in Visual Studio 2019 version 16.6. To get the new version that includes the fix, [check for the latest updates](https://docs.microsoft.com/visualstudio/install/update-visual-studio?view=vs-2019) or install the latest version from <https://visualstudio.microsoft.com/downloads/>.

New functionality included on macOS in Visual Studio 2019 for Mac version 8.6. To get the new version that includes the fix, check for the latest updates on the **Stable** [updater channel](https://docs.microsoft.com/visualstudio/mac/update).
